devlib/utils/ssh.py

```diff
@@ -170,51 +305,422 @@ def name(self):
+
+        # If there is anything we need to redirect to, spawn a thread taking
+        # care of that
+        select_timeout = 1
+        thread_out_streams = {
+            name: (r, w)
+            for name, (r, w) in out_streams.items()
+            if w is not None
+        }
+        redirect_thread = threading.Thread(
+            target=redirect_thread_f,
+            args=(stdout_in, stderr_in, thread_out_streams, select_timeout),
+            # The thread will die when the main thread dies
+            daemon=True,
+        )
+        redirect_thread.start()
+
+        return ParamikoBackgroundCommand(
+            conn=self,
+            as_root=as_root,
+            chan=channel,
+            pid=pid,
+            stdin=stdin,
+            # We give the reading end to the consumer of the data
+            stdout=out_streams['stdout'][0],
+            stderr=out_streams['stderr'][0],
+            redirect_thread=redirect_thread,
+        )
+
+    def _close(self):
+        logger.debug('Logging out {}@{}'.format(self.username, self.host))
+        with _handle_paramiko_exceptions():
+            bg_cmds = set(self._current_bg_cmds)
+            for bg_cmd in bg_cmds:
+                bg_cmd.close()
+            self.client.close()
+
+    def _execute_with_sudo(self, command, as_root, log, timeout, executor):
+        # As we're already root, there is no need to use sudo.
+        log_debug = logger.debug if log else lambda msg: None
+        use_sudo = as_root and not self.connected_as_root
+
+        if use_sudo:
+            if not self.password:
+                raise TargetStableError('Attempt to use sudo but no password was specified')
+
+            command = self.sudo_cmd.format(quote(command))
+
+            log_debug(command)
+            streams = executor(command, timeout=timeout)
+            stdin = streams[0]
+            stdin.write(self.password + '\n')
+            stdin.flush()
+        else:
+            log_debug(command)
+            streams = executor(command, timeout=timeout)
+
+        return streams
+
+    def _execute(self, command, timeout=None, as_root=False, strip_colors=True, log=True):
+        # Merge stderr into stdout since we are going without a TTY
+        command = '({}) 2>&1'.format(command)
+
+        stdin, stdout, stderr = self._execute_with_sudo(
+            command,
+            as_root=as_root,
+            log=log,
+            timeout=timeout,
+            executor=self.client.exec_command,
+        )
+        stdin.close()
+
+        # Empty the stdout buffer of the command, allowing it to carry on to
+        # completion
+        def callback(output_chunks, name, chunk):
+            output_chunks.append(chunk)
+            return output_chunks
+
+        select_timeout = 1
+        output_chunks, exit_code = _read_paramiko_streams(stdout, stderr, select_timeout, callback, [])
+        # Join in one go to avoid O(N^2) concatenation
+        output = b''.join(output_chunks)
```

Comment:
The streams from the command itself are actually merged by:
```

        # Merge stderr into stdout since we are going without a TTY
        command = '({}) 2>&1'.format(command) 
```
The setup you have seems to have flaky configuration, I'm not sure it's a good idea to get rid of stderr on some arbitrary criteria like return code == 0. Even without considering that, this behavior would be impossible to implement on adb, meaning any code written against an SSH connection relying on that would stop working on some other connection types.

For now, the `execute()` method merges stdout and stderr, if something else is needed `background()` can be used for more manual process control. It does allow separate handling of stdout/stderr.

That said we could add a "discard_stderr=False" option to `execute()`, which should be implementable for all connection types. It won't swallow shell config problems like you have (unless we wrap everything with a shell command that does not read any config and that discards stderr).

Reply:
Agreed that my setup is not quite correctly configured however in this case I don't think the std error is coming from the command invocation and therefore is not merged initially, this is coming from the terminal itself, as you can see we're getting separate `output_chunks`.

I'm sure also not sure that disregarding it would make sense or what the behaviour should be. Just wanted to highlight the behaviour I'd noticed so far is unclear, this manifested as a `FileNotFound` error when attempting to use a path discovered from the device (since the path also included this error message).

Reply:
The same would happen if e.g. a command is not found. There is no way to tell the difference between something written on stderr by the shell and something written by the command you care about. The way most APIs like `subprocess` handle that is by making stderr opt-in, which is usually what people care about. In devlib case, `execute()` merges both as things like adb don't support splitting the two.

The only way of "generically" fixing that problem is changing `execute()` semantic to discard stderr by default, which is an API break but may actually improve reliability of most existing code.

Reply:
That makes sense and I think you're right, breaking the API or trying to hide the output arbitrarily probably isn't the best idea. I guess if possible the underlying problem with the environment should just be fixed (simple enough for my example case) and if not then it could be dealt with explicitly for the particular device.